fiftyone/utils/beam.py

```diff
@@ -508,6 +703,135 @@ def process(self, element, **kwargs):
+
+        dataset = fo.load_dataset(self.dataset_name)
+
+        if self.view_stages:
+            sample_collection = fov.DatasetView._build(
+                dataset, self.view_stages
+            )
+        else:
+            sample_collection = dataset
+
+        self._sample_collection = sample_collection
+
+    def process(self, element, **kwargs):
+        i, num_batches, batch = element
+
+        if isinstance(batch, tuple):
+            # Slice batches
+            start, stop = batch
+            total = stop - start
+            batch_view = self._sample_collection[start:stop]
+        else:
+            # ID batches
+            sample_ids = batch
+            total = len(sample_ids)
+            batch_view = fov.make_optimized_select_view(
+                self._sample_collection, sample_ids
+            )
+
+        if self.progress:
+            desc = f"Batch {i + 1:0{len(str(num_batches))}}/{num_batches}"
+            with tqdm.tqdm(total=total, desc=desc, position=i) as pbar:
+                for sample in batch_view.iter_samples(autosave=self.save):
+                    result = self.map_fcn(sample)
+                    if self.return_outputs and result is not None:
+                        yield sample.id, result
+
+                    pbar.update()
+        else:
+            for sample in batch_view.iter_samples(autosave=self.save):
+                result = self.map_fcn(sample)
+                if self.return_outputs and result is not None:
+                    yield sample.id, result
+
+
+class ReduceFn(beam.CombineFn):
+    def __init__(
+        self,
+        dataset_name,
+        reduce_fcn,
+        view_stages=None,
+        result_key=None,
+    ):
+        self.dataset_name = dataset_name
+        self.reduce_fcn = reduce_fcn
+        self.view_stages = view_stages
+        self.result_key = result_key
+
+        self._sample_collection = None
+
+    def setup(self):
+        import fiftyone as fo
+        import fiftyone.core.view as fov
+
+        dataset = fo.load_dataset(self.dataset_name)
+
+        if self.view_stages:
+            sample_collection = fov.DatasetView._build(
+                dataset, self.view_stages
+            )
+        else:
+            sample_collection = dataset
+
+        self._sample_collection = sample_collection
+
+    def create_accumulator(self):
+        return {}
+
+    def add_input(self, accumulator, input):
+        sample_id, value = input
+        accumulator[sample_id] = value
+        return accumulator
+
+    def merge_accumulators(self, accumulators):
+        accumulator = {}
+        for a in accumulators:
+            accumulator.update(a)
+        return accumulator
+
+    def extract_output(self, accumulator):
+        output = self.reduce_fcn(self._sample_collection, accumulator)
+        if output is None or self.result_key is None:
+            return
+
+        _set_key(self._sample_collection, self.result_key, output)
```

Comment:
do we need to return `output` here so that the output from the return function is returned?

Reply:
No, the output is stored via `_set_key()` and loaded via `_get_key()` back in the main method.

Apache Beam supports distributed computing, so in general `ReduceFn.extract_output()` may be called on a different machine and thus any I/O must be done via disk, database, etc.